File: Review-06-Java-Shapes/Example-7/src/main/java/edu/odu/cs/cs330/examples/shapes/Shape.java
// Thomas Kennedy

package edu.odu.cs.cs330.examples.shapes;

/**
 * Shape in a 2-D Cartesian Plane.
 *
 * @author Thomas J Kennedy
 */
public abstract class Shape implements Cloneable {

    /**
     * Base for Shape formatting strings.
     */
    protected static final String FMT_BASE = "%-12s: %24";

    /**
     * The string-value format string.
     */
    protected static final String FMT_STR = FMT_BASE + "s%n";

    /**
     * The double-value format string.
     */
    protected static final String FMT_DBL = FMT_BASE + ".4f%n";

    /**
     * Return the name.
     *
     * @return shape name
     */
    public abstract String name();

    /**
     * Compute the area.
     *
     * @return area
     */
    public abstract double area();

    /**
     * Compute the perimeter.
     *
     * @return perimeter
     */
    public abstract double perimeter();

    /**
     * Return a new duplicate Shape.
     */
    public abstract Object clone() throws CloneNotSupportedException;

    @Override
    public String toString()
    {
        return String.format(FMT_STR, "Name", this.name());
    }
}

File: Review-06-Rust-Shapes/Example-3/src/shape.rs
use std::fmt::Debug;
use std::fmt::Display;

/// Define the interface for a 2D shape
pub trait Shape: Display + Debug {
    /// Return the name of the shape as a fixed string
    fn name(&self) -> &'static str;

    /// Compute the area of a 2D shape
    fn area(&self) -> f64;

    /// Compute the perimeter of a 2D shape
    fn perimeter(&self) -> f64;
}
